### ISSUE TITLE
Tag canvas visuals

### ISSUE THREAD

**Issue by ferdnyc** [BiglySoftware/BiglyBT#1221]
I like the _idea_ of the tag canvas, to replace the clunky old buttons, but I'm not sure it's _quite_ at "ALL CAPS OMG!" levels just yet. :wink: 

Here's a before/after/after-after for my test install, after upgrading to _B29 (all three images at the same scale):

![Screenshot from 2019-08-30 13-55-43](https://user-images.githubusercontent.com/538020/64042948-0dfc8600-cb31-11e9-8a8d-3066c52936bb.png)
![Screenshot from 2019-08-30 14-01-21](https://user-images.githubusercontent.com/538020/64042958-1785ee00-cb31-11e9-9852-c72f76022ce8.png)
![Screenshot from 2019-08-30 14-05-29](https://user-images.githubusercontent.com/538020/64042972-240a4680-cb31-11e9-9e04-511d54c72010.png)

* First off, there's no delicate way to say this: Solid colors? Great! But that dashed selection outline looks like _ass_.

* Second, making the size dependent on the label might require a minimum width or something, because really short tag names are suddenly a problem. (The buttons had a large fixed minimum width.) I have one tag with a two-letter name, and it practically requires tweezers now. :laughing: 

* And, a bug I just discovered: Color and name changes in Tags Overview don't propagate to the canvas view automatically (even though they're visible in the sidebar).

* Finally, while the canvas isn't too bad under Gnome Shell, here's what it looks like in my main LXDE setup (with most of the names obscured for privacy):
    ![Screenshot from 2019-08-30 14-06-31](https://user-images.githubusercontent.com/538020/64043206-c5919800-cb31-11e9-899e-0c2e19b42c08.png)
    Yes, the last "s" in the third tag's name really is slightly _over_ the border. I don't know why the two desktops render so differently, whether it's down to interface font choice or what. Even the height is squished, which just leaves random extra blank space on the toolbar underneath.

As I said, I actually really like the idea! Lots of potential. A few tweaks to the execution and it can live up to it.

#### System Details
Various Linux, various BiglyBT versions (but mosly 2.0.0.1_B29).

**Comment by TuxPaper** [comment]
Thanks for the feedback :)

I'm not totally happy with the focus ring yet either (actually, the focused, unselected tag rendering is what bothers me) so expect some changes eventually.  I might make it a 1px line, 1px inside the tag (so there'd be a 1 px solid line around still), but I'm not sure how that will look yet.

Minimum width shouldn't be a problem, I'll add something real soon.

Color changing showing up was working yesterday, so I probably just broke something.

Linux sucks!  I mean, ooh, LXDE, how pretty. I'll make a note to fiddle with it on LXDE, but won't get to that until the tag drawing is more finalized.

**Comment by ferdnyc** [comment]
I see a bunch of code for "compact" rendering, maybe that's what it's doing under LXDE? I can't seem to find the code that _drives_ that decision, though. What ultimately controls whether a tag button is  drawn compact? Is that a standard SWT flag, or something in the BiglyBT implementation? (If the former, maybe the fact that my LXDE desktop is only a 1280x720 Xvnc display has something to do with it?)

**Comment by ferdnyc** [comment]
> I'm not totally happy with the focus ring yet either (actually, the focused, unselected tag rendering is what bothers me)

Not to pile on, but since you mentioned it (TBH, I hadn't even realized that a tag button _could_ be focused unselected! But, you just _had_ to go and poke my inner QA bear.) 

...since you mentioned it, :angel: it seems that when an unselected tag button is focused, neither <kbd>Space</kbd> nor <kbd>Enter</kbd> works to select it.

**Comment by TuxPaper** [comment]
Next beta will have a different focus indicator, the colors should update, min tag width of something slightly larger than a sliver, space/enter should be working, and that gap a the bottom of the tag bar should be gone.  Not sure if the font thing is fixed for LXDE.

I noticed that adding a tag doesn't always show it in the tagging view until a hard refresh, but I'll have to fixed that next time.

**Comment by ferdnyc** [comment]
Wow, just updated from B30 to B33, that's the biggest jump in forever! :laughing: 

> * Next beta will have a different focus indicator,
> * min tag width of something slightly larger than a sliver,
> * space/enter should be working,
> * and that gap a the bottom of the tag bar should be gone.
> * Not sure if the font thing is fixed for LXDE.

Check, check, check, check, and it is! Fonts fit completely within the tag border. The focus indicator hugs it pretty tightly (it's literally the line of pixels directly above the bottom border), which makes it a little hard to spot with darker tag colors if the tag is unselected (so the dashes are black, not white), but that's really NBD.

Only this one is a nope:
> the colors should update,

Which I assume is related to this: 
> I noticed that adding a tag doesn't always show it in the tagging view until a hard refresh, but I'll have to fixed that next time.

Editing color or even tag _name_ isn't reflected in the canvas after closing, though it is in the Tags column and sidebar.

This is the state long after I closed Tags Overview (more than an hour later, in fact ­— I got distracted). The color should match the first one in the Tags column. I'd also edited the tag name, which didn't change.

![image](https://user-images.githubusercontent.com/538020/64074065-3640cd80-cc74-11e9-83c0-c9e86c41a367.png)
